src/translators/crate.py

```diff
@@ -550,41 +615,109 @@ def query(self,
+                entities = self._format_response(res,
+                                                 col_names,
+                                                 tn,
+                                                 last_n,
+                                                 aggr_method)
             result.extend(entities)
-
-        if last_n:
-            # TODO: embed last_n in query to avoid waste.
-            return result[-last_n:]
         return result
 
-    # TODO: Remove this method (needs refactoring of the benchmark)
-    def average(self, attr_name, entity_type=None, entity_id=None):
-        if entity_id and not entity_type:
-            msg = "For now you must specify entity_type when stating entity_id"
-            raise NotImplementedError(msg)
 
-        if entity_type:
-            table_names = [self._et2tn(entity_type)]
-        else:
-            # The semantic correctness of this operation is up to the client.
-            table_names = self._get_et_table_names()
+    def _format_response(self, resultset, col_names, table_name, last_n,
+                         aggr_Method):
+        """
+        :param resultset: list of query results for one entity_type
+        :param col_names: list of columns affected in the query
+        :param table_name: name of table where the query took place.
+        :param last_n: see last_n in query method.
+        :param aggr_method: True if used in the request, false otherwise.
+
+        :return: list of dicts. Possible scenarios
+
+        Without aggrMethod, there will be one dict per entity instance. E.g.,
+        returns [
+                {'type': 'Room',
+                'id': 'Room1',
+                'temperature': {'values': [1, 2, 3]}
+                'index: [t0, t1, t2]
+                },
+                {'type': 'Room',
+                'id': 'Room2',
+                'temperature': {'values': [5, 6, 4]}
+                'index: [t0, t1, t2]
+                }
+                ]
+
+        With aggrMethod and one specific id requested, the list has only one
+        dict. The number of values is 1 if no aggrPeriod was asked for,
+        or one value per aggregation period step.
+        returns [{
+                    'type': 'Room',
+                    'id': 'SpecificRoom',
+                    'temperature': {'values': [1, 2, 3]},   # aggrMethod
+                    'index': [tA, tB, tC]                   # with aggrPeriod
+                }]
+
+        With aggrMethod and many specific ids requested, the array has
+        only one dict, but instead of an 'id' attribute it will have an
+        'ids'. This is the same case when user does not give a specific id at
+        all (no id => all ids).
+        returns [{
+                    'type': 'Room',
+                    'ids': ['Room1', 'Room2'],
+                    'temperature': {'values': [4,]},    # aggrMethod
+                    'index': [],                        # without aggrPeriod
```

Comment:
why not index?

Reply:
it's the case where user asks something like: 'average temperature for entities 1&2 from 2010 to 2011'. The response is just a number. There's no actual 'index'.

Another option is to make it an array of 2 steps: fromDate and toDate. That'd be the idea we mentioned in [this unimplemented method](https://github.com/smartsdk/ngsi-timeseries-api/blob/bc8d0bcb97b3e3d2709649cc34d7add9240a1d4c/specification/quantumleap.yml#L849). I'm still not convinced of its added value / effort ratio though.

Reply:
fair enough